wagtail/core/tests/test_sites.py

```diff
@@ -56,39 +58,207 @@ def test_no_site_name_default_port_n(self):
-            request = HttpRequest()
-            request.META = {'HTTP_X_FORWARDED_HOST': 'example.com'}
-            self.assertEqual(Site.find_for_request(request), self.site)
+        request = HttpRequest()
+        request.META = {'HTTP_X_FORWARDED_HOST': 'about.example.com'}
+        self.assertEqual(Site.find_for_request(request), self.about_site)
+
+    def test_no_host_header_routes_to_default_site(self):
+        # requests without a Host: header should be directed to the default site
+        request = HttpRequest()
+        request.path = '/'
+        with self.assertNumQueries(self.find_for_request_queries_expected):
+            self.assertEqual(Site.find_for_request(request), self.default_site)
+
+    def test_valid_headers_route_to_specific_site(self):
+        # requests with a known Host: header should be directed to the specific site
+        request = HttpRequest()
+        request.path = '/'
+        request.META['HTTP_HOST'] = self.events_site.hostname
+        request.META['SERVER_PORT'] = self.events_site.port
+        with self.assertNumQueries(self.find_for_request_queries_expected):
+            self.assertEqual(Site.find_for_request(request), self.events_site)
+
+    def test_ports_in_request_headers_are_respected(self):
+        # ports in the Host: header should be respected
+        request = HttpRequest()
+        request.path = '/'
+        request.META['HTTP_HOST'] = self.alternate_port_events_site.hostname
+        request.META['SERVER_PORT'] = self.alternate_port_events_site.port
+        with self.assertNumQueries(self.find_for_request_queries_expected):
+            self.assertEqual(Site.find_for_request(request), self.alternate_port_events_site)
+
+    def test_unrecognised_host_header_routes_to_default_site(self):
+        # requests with an unrecognised Host: header should be directed to the default site
+        request = HttpRequest()
+        request.path = '/'
+        request.META['HTTP_HOST'] = self.unrecognised_hostname
+        request.META['SERVER_PORT'] = '80'
+        with self.assertNumQueries(self.find_for_request_queries_expected):
+            self.assertEqual(Site.find_for_request(request), self.default_site)
+
+    def test_unrecognised_port_and_default_host_routes_to_default_site(self):
+        # requests to the default host on an unrecognised port should be directed to the default site
+        request = HttpRequest()
+        request.path = '/'
+        request.META['HTTP_HOST'] = self.default_site.hostname
+        request.META['SERVER_PORT'] = self.unrecognised_port
+        with self.assertNumQueries(self.find_for_request_queries_expected):
+            self.assertEqual(Site.find_for_request(request), self.default_site)
+
+    def test_unrecognised_port_and_unrecognised_host_routes_to_default_site(self):
+        # requests with an unrecognised Host: header _and_ an unrecognised port
+        # hould be directed to the default site
```

Comment:
See comment above; in the existing logic, isn't this only true if there's only one site, otherwise it should raise `Site.DoesNotExist`?

Reply:
This test was copied pretty much 'as is' from [`test_page_models.py`](https://github.com/wagtail/wagtail/pull/4142/files/7e1bc7c06bd52210616d4010f5024c46236bb0fd#diff-04b4575a8f6981406bfdca2128f7a88bL177), so I'm sure the behaviour described by the test is valid. However, I will add a new test to confirm that Site.DoesNotExist is raised when no site is marked as the default (I'm pretty sure that's the only time you'd run into that exception).